### ISSUE TITLE
[Bug?] Slow polling performance 

### ISSUE THREAD

**Issue by pavleec** [toni-moreno/snmpcollector#153]
Hi,

I've just started using snmpcollector which looks very promising but it's polling performance is very low at the moment. I have only one host defined (SNMP v2c, interval 60s, bulks enabled) with 86 metrics and the polling tooks usually 10-12s which is terrible. The same polling using snmpbulkwalk commad tooks less then <2s. Any plans to improve that?

```
Statistics SNMP Device
Gets 86
Errors 0
Last Gather Duration 10.886523241
Last Filter Duration 0.230377323
```

**Comment by toni-moreno** [comment]
Performance are typically better than you are getting , are you sure you have bulk enabled? enabled is the default configuration and you should'n change it.

Anyway  Let me see your complete configuration  ( screenshots could be enough) , Log Level to debug  ( you can do online from the runtime options) 

![image](https://cloud.githubusercontent.com/assets/5883405/23950362/07437202-098b-11e7-8ac9-e579340d3da1.png)

And send us the generated file in  logs/<device_id>.log you can also send us the generic log in logs/snmpcollector.log

**Comment by pavleec** [comment]
Ok, here is my device configuration with snmpbulk enabled:

![mcc](https://cloud.githubusercontent.com/assets/12551929/23955135/5e39bd88-0999-11e7-937f-253f8765ecde.jpg)

For some OIDs I'm also using STRINGPARSER ([0-9]+)  because the're actual INTEGERs but returned as STRINGs.

![parser](https://cloud.githubusercontent.com/assets/12551929/23955481/488ddd42-099a-11e7-8fde-4f1fd3d79574.png)

And here is the device DEBUG log file:
[host-a-1.zip](https://github.com/toni-moreno/snmpcollector/files/844976/host-a-1.zip)

**Comment by toni-moreno** [comment]
Hi @pavleec.

SnmpCollector right now works with one "thread" for device , and data gathering is done sequentially . When taking data from a measurement , snmpcollector does a snmpbulkwalk for each snmpmetric defined in the measurement and  measurements are gathered in a sequential loop 

 I'm reviewing you log file. And I can see you are working with 4 measurements , with this number of fields

* mcc_networkcontext (26 fields) - defined as table
* mcc_radius (8 fields)- defined as table
* mcc_httpproxy (51 fields)- definded as table
* mcc_callperf(34 fields) -defined as table

If I'm not wrong you are doing 119 sequential snmpbulkget in 20 seconds... I think is good enough, is'n it ?

Anyway we have plans at least to improve this question adding  a parallel thread for each measurement , but we don't know exactly when. Sorry

**Comment by pavleec** [comment]
I came across your project when I was googling for a beter solution to collect SNMP statistics and write them to InfluxDB. I've tried to use Telegraf snmp input so far but, as someone has already mentioned here it's light years behind. However I'm still using a slightly modified version of Cacti which outputs (every 30s) the data collected by it's spine poller not only to RRD files but also to Graphite (Whisper DB). It's running on the same host as snmpcollector and  statistics looks like below:

`Time 14.4254 Method:spine Processes:1 Threads:10 Hosts:95 HostsPerProcess:95 DataSources:12244 RRDsProcessed:9902 `

That is about 880 snmpbulksget per second or 87 gets per one thread. Spine poller code is probably fully optimzed. 
You've just started so I really support your project and keep fingers crossed for it's evolving success :)

**Comment by toni-moreno** [comment]
@pavleec snmpcollector works with one "thread" per device , and it will be able to get data from hundreds of  devices.  Each "thread" is like a poller itself while the cacti poller spine works in a very different way as a only entity to get all data.

Performance depens on the device and also on how measurements and fields are defined , as you can see in the next picture we are getting data from cisco switches with hundreds of ports in a few seconds.

![image](https://cloud.githubusercontent.com/assets/5883405/24017716/651c21a6-0a91-11e7-9c44-6c854ddc7760.png)

this one is getting 5870 metrics in 11 seconds (,5 Measurements with (13+18+3+1+1) = 36 Fields. ) . 

And you would be able to get data from dozens of switches like this one gathering data each one in its dedicated "thread". (we have plans to deploy one snmpcollector agent getting data from 500 switches like that in next weeks )

Anyway we will try to improve performance by doing parallel gathering for each measurement in each device.

**Comment by pavleec** [comment]
Thanks for a detailed explanation. I will try to move all of the the Cacti statistics and let you know how it works then. Polling performance is one thing, but the architecture of snmpcollector and it's simplicity of adding new metrics makes it the winner in this case. I'm looking forward for the new releases.

**Comment by toni-moreno** [comment]
Hi @pavleec new released version 0.6.6 has new improvements focused on performance.

1) it has been compiled with new golang 1.8 (with multiple improvements)

2) has a new feature "Concurrent Gather" ( enabled by default) that let you send concurrent queries over devices as many as measurements you have configured in your device. As we could see in our tests this new feature not always has the expected behaviour --it depends on the device and how it handles concurrent snmp queries-- but it usually let us reduce  20%/30% the total gather time.

3) Enables fine tuning for "MaxRepetitions"   that can increase big snmp walks performance (and also 
You can also fix some problems if the default MaxRepetiion = 50 is not supported on your devices a in https://github.com/toni-moreno/snmpcollector/issues/161) , also can crash the snmpwalk response if using greater values than device can support.(be carefully with it) 

4) we have also Updated the main snmp library (gosnmp) with a big snmp v3 performance improvement.
 
I hope this should be enough for now.  Feel free to share with us if you could see some improvements
